src/coreclr/jit/lclvars.cpp

```diff
@@ -1405,6 +1429,33 @@ bool Compiler::lvaInitSpecialSwiftParam(InitVarDscInfo* varDscInfo, CorInfoType
         return true;
     }
 
+    if ((strcmp(className, "SwiftError") == 0) && (strcmp(namespaceName, "System.Runtime.InteropServices.Swift") == 0))
+    {
+        if (!argIsByrefOrPtr)
+        {
+            BADCODE("Expected SwiftError pointer/reference, got struct");
+        }
+
+        if (lvaSwiftErrorArg != BAD_VAR_NUM)
+        {
+            BADCODE("Duplicate SwiftError* parameter");
+        }
+
+        // We won't actually be passing this SwiftError* in REG_SWIFT_ERROR (or any register, for that matter).
+        // We will check for this quirk when generating the prolog,
+        // and ensure this fake parameter doesn't take any registers/stack space
+        LclVarDsc* const varDsc = varDscInfo->varDsc;
+        varDsc->SetArgReg(REG_SWIFT_ERROR);
+        varDsc->SetOtherArgReg(REG_NA);
+        varDsc->lvIsRegArg = true;
+        lvaSwiftErrorArg   = varDscInfo->varNum;
+
+        // Instead, all usages of the SwiftError* parameter will be redirected to this pseudolocal.
+        lvaSwiftErrorLocal = lvaGrabTempWithImplicitUse(false DEBUGARG("SwiftError pseudolocal"));
```

Comment:
Can you verify that something somewhere marks this local as either DNER or address exposed? I don't immediately see where we do that in the JIT when something has implicit uses. Basically, I don't understand for example why liveness doesn't get rid of stores to this local because it cannot see any uses of it. Can you share a jitdump of one of the tests just so I can understand why that doesn't happen?

Reply:
I think we are getting "lucky" there. The JIT is spilling the address of the local to a temp (not really very optimal, we ought to improve this). It only does that because of the constructor call on the RHS of the assignment, and as a consequence we end up address exposing it. If you changed the code slightly, for example to
```csharp
var x = new SwiftError(...);
GC.KeepAlive(null); // avoid Roslyn optimizing it back to the pattern
*error = x;
```
then I would expect that you will hit the issue I describe where nothing marks it as address exposed and we end up removing the store.

Reply:
I tried this, and you're correct that we no longer mark it as address exposed, but we don't end up removing the store -- I think this is because we mark all `GT_LCL_ADDR` nodes pointing to the pseudolocal as having side effects (perhaps that's what is forcing the JIT to spill the address to a temp?). I'm fine with going back to explicitly marking the pseudolocal as address exposed upon creation. My hope is that is short-lived anyway, once I open a PR for `GT_SWIFT_ERROR_RET`.

Reply:
> I'm fine with going back to explicitly marking the pseudolocal as address exposed upon creation

This, along with removing the side effects on the `GT_LCL_ADDR` nodes of the pseudolocal, improves codegen. For example, before:

```
; Assembly listing for method ErrorHandlingTests:ConditionallySetErrorTo21(ulong,int) (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Unix
; FullOpts code
; optimized code
; rbp based frame
; partially interruptible
; No PGO data
; 0 inlinees with PGO data; 2 single block inlinees; 0 inlinees without PGO data
; Final local variable assignments
;
;* V00 arg0         [V00    ] (  0,  0   )    long  ->  zero-ref    single-def
;  V01 arg1         [V01,T00] (  3,  3   )     int  ->  rbx         single-def
;* V02 loc0         [V02    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op <System.Runtime.InteropServices.Swift.SwiftError>
;  V03 tmp0         [V03    ] (  3,  2   )  struct ( 8) [rbp-0x10]  do-not-enreg[XS] must-init addr-exposed "SwiftError pseudolocal" <System.Runtime.InteropServices.Swift.SwiftError>
;# V04 OutArgs      [V04    ] (  1,  1   )  struct ( 0) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;* V05 tmp2         [V05    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op "NewObj constructor temp" <System.Runtime.InteropServices.Swift.SwiftError>
;  V06 tmp3         [V06,T01] (  2,  2   )    long  ->  rdi         "impAppendStmt"
;  V07 tmp4         [V07    ] (  3,  3   )  struct (16) [rbp-0x20]  do-not-enreg[XS] must-init addr-exposed "Reverse Pinvoke FrameVar"
;* V08 tmp5         [V08,T02] (  0,  0   )    long  ->  zero-ref    single-def "field V02.<Value>k__BackingField (fldOffset=0x0)" P-INDEP
;  V09 tmp6         [V09    ] (  2,  1   )    long  ->  [rbp-0x10]  do-not-enreg[X] addr-exposed "field V03.<Value>k__BackingField (fldOffset=0x0)" P-DEP
;* V10 tmp7         [V10,T03] (  0,  0   )    long  ->  zero-ref    single-def "field V05.<Value>k__BackingField (fldOffset=0x0)" P-INDEP
;
; Lcl frame size = 24

G_M3386_IG01:  ;; offset=0x0000
       push     rbp
       push     rbx
       sub      rsp, 24
       lea      rbp, [rsp+0x20]
       vxorps   xmm8, xmm8, xmm8
       vmovdqa  xmmword ptr [rbp-0x20], xmm8
       xor      eax, eax
       mov      qword ptr [rbp-0x10], rax
       mov      ebx, edi
						;; size=29 bbWeight=1 PerfScore 6.58
G_M3386_IG02:  ;; offset=0x001D
       lea      rdi, [rbp-0x20]
       call     CORINFO_HELP_JIT_REVERSE_PINVOKE_ENTER
       test     ebx, ebx
       je       SHORT G_M3386_IG04
						;; size=13 bbWeight=1 PerfScore 2.75
G_M3386_IG03:  ;; offset=0x002A
       xor      rdi, rdi
       mov      qword ptr [rbp-0x10], 21
       jmp      SHORT G_M3386_IG05
						;; size=12 bbWeight=0.50 PerfScore 1.62
G_M3386_IG04:  ;; offset=0x0036
       lea      rdi, [rbp-0x10]
       xor      eax, eax
       mov      qword ptr [rdi], rax
						;; size=9 bbWeight=0.50 PerfScore 0.88
G_M3386_IG05:  ;; offset=0x003F
       lea      rdi, [rbp-0x20]
       call     CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT
       mov      r12, qword ptr [rbp-0x10]
						;; size=13 bbWeight=1 PerfScore 2.50
G_M3386_IG06:  ;; offset=0x004C
       add      rsp, 24
       pop      rbx
       pop      rbp
       ret      
						;; size=7 bbWeight=1 PerfScore 2.25

; Total bytes of code 83, prolog size 27, PerfScore 16.58, instruction count 26, allocated bytes for code 83 (MethodHash=763af2c5) for method ErrorHandlingTests:ConditionallySetErrorTo21(ulong,int) (FullOpts)
; ============================================================
```

After:
```
; Assembly listing for method ErrorHandlingTests:ConditionallySetErrorTo21(ulong,int) (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Unix
; FullOpts code
; optimized code
; rbp based frame
; partially interruptible
; No PGO data
; 0 inlinees with PGO data; 2 single block inlinees; 0 inlinees without PGO data
; Final local variable assignments
;
;* V00 arg0         [V00    ] (  0,  0   )    long  ->  zero-ref    single-def
;  V01 arg1         [V01,T00] (  3,  3   )     int  ->  rbx         single-def
;* V02 loc0         [V02    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op <System.Runtime.InteropServices.Swift.SwiftError>
;  V03 tmp0         [V03    ] (  3,  2   )  struct ( 8) [rbp-0x10]  do-not-enreg[XS] must-init addr-exposed "SwiftError pseudolocal" <System.Runtime.InteropServices.Swift.SwiftError>
;# V04 OutArgs      [V04    ] (  1,  1   )  struct ( 0) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;* V05 tmp2         [V05    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op "NewObj constructor temp" <System.Runtime.InteropServices.Swift.SwiftError>
;  V06 tmp3         [V06    ] (  3,  3   )  struct (16) [rbp-0x20]  do-not-enreg[XS] must-init addr-exposed "Reverse Pinvoke FrameVar"
;* V07 tmp4         [V07,T01] (  0,  0   )    long  ->  zero-ref    single-def "field V02.<Value>k__BackingField (fldOffset=0x0)" P-INDEP
;* V08 tmp5         [V08,T02] (  0,  0   )    long  ->  zero-ref    single-def "field V05.<Value>k__BackingField (fldOffset=0x0)" P-INDEP
;
; Lcl frame size = 24

G_M3386_IG01:  ;; offset=0x0000
       push     rbp
       push     rbx
       sub      rsp, 24
       lea      rbp, [rsp+0x20]
       vxorps   xmm8, xmm8, xmm8
       vmovdqa  xmmword ptr [rbp-0x20], xmm8
       xor      eax, eax
       mov      qword ptr [rbp-0x10], rax
       mov      ebx, edi
						;; size=29 bbWeight=1 PerfScore 6.58
G_M3386_IG02:  ;; offset=0x001D
       lea      rdi, [rbp-0x20]
       call     CORINFO_HELP_JIT_REVERSE_PINVOKE_ENTER
       test     ebx, ebx
       je       SHORT G_M3386_IG04
						;; size=13 bbWeight=1 PerfScore 2.75
G_M3386_IG03:  ;; offset=0x002A
       xor      rdi, rdi
       mov      qword ptr [rbp-0x10], 21
       jmp      SHORT G_M3386_IG05
						;; size=12 bbWeight=0.50 PerfScore 1.62
G_M3386_IG04:  ;; offset=0x0036
       xor      edi, edi
       mov      qword ptr [rbp-0x10], rdi
						;; size=6 bbWeight=0.50 PerfScore 0.62
G_M3386_IG05:  ;; offset=0x003C
       lea      rdi, [rbp-0x20]
       call     CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT
       mov      r12, qword ptr [rbp-0x10]
						;; size=13 bbWeight=1 PerfScore 2.50
G_M3386_IG06:  ;; offset=0x0049
       add      rsp, 24
       pop      rbx
       pop      rbp
       ret      
						;; size=7 bbWeight=1 PerfScore 2.25

; Total bytes of code 80, prolog size 27, PerfScore 16.33, instruction count 25, allocated bytes for code 80 (MethodHash=763af2c5) for method ErrorHandlingTests:ConditionallySetErrorTo21(ulong,int) (FullOpts)
; ============================================================
```

Reply:
Ah, that makes sense. Yeah, the side effect flags on the `GT_LCL_ADDR`  shouldn't be necessary, 

> My hope is that is short-lived anyway, once I open a PR for `GT_SWIFT_ERROR_RET`.

Do you plan to do the work? To be honest looking at the codegen there it really doesn't feel all that bad compared to all the other things we end up doing for the reverse pinvokes anyway (like the helper calls), so I'd be perfectly fine with just leaving it like this.

Reply:
I have some of the implementation saved locally, so I'm interested in trying it out; I think it can be done without polluting the JIT with too many weird edge cases.